kwasm-lemmas.md

```diff
@@ -308,68 +310,51 @@ We want to make the variant explicit, so we introduce the following helper, whic
       requires notBool isInt(I)
 ```
 
-We can now state the intended semantics of `#get`.
-
-```k
-    rule LOW            <=Int #get(BM, ADDR) => true
-      requires LOW <=Int 0
-       andBool #isByteMap(BM) [simplification]
-
-    rule #get(BM, ADDR)  <Int HIGH           => true
-      requires 256 <=Int HIGH
-       andBool #isByteMap(BM) [simplification]
-```
-
 With this invariant encoded, we can introduce the following lemma.
 
 ```k
     rule #isByteMap(BMAP) impliesBool (0 <=Int #get(BMAP, IDX) andBool #get(BMAP, IDX) <Int 256) => true [smt-lemma]
 ```
 
-From the semantics, it should be clear that setting the index in a bytemap to the value already contained there will leave the map unchanged.
-Conversely, setting an index in a map to a value `VAL` and then retrieving the value at that index will yield `VAL`.
-A getting operation can safeley ignore setting operation on unrelated indices.
-
-```k
-    rule #set(BMAP, IDX, #get(BMAP, IDX)) => BMAP [smt-lemma]
-
-    rule #get(#set(BMAP, IDX', VAL), IDX) => VAL             requires         IDX ==Int IDX'
-    rule #get(#set(BMAP, IDX', VAL), IDX) => #get(BMAP, IDX) requires notBool IDX ==Int IDX'
-```
-
 TODO: We should inspect the two functions `#getRange` and `#setRange` closer.
 They are non-trivial in their implementation, but the following should obviously hold from the intended semantics.
 
 ```k
-    rule #setRange(BM, EA, #getRange(BM, EA, WIDTH), WIDTH) => BM
-    rule #setRange(BM, EA, VAL, WIDTH) => #set(BM, EA, #wrap(1, VAL))
-      requires WIDTH ==Int 1
-      [simplification]
-
-    rule #getRange(BM, ADDR, WIDTH) modInt BYTE_SIZE => #get(BM, ADDR)
-      requires BYTE_SIZE ==Int 256
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) >>Int SHIFT => #getRange(BM, ADDR +Int 1, WIDTH -Int 1) >>Int (SHIFT -Int 8)
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool 8 <=Int SHIFT
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #get(BM, ADDR)
-      requires N ==Int 1
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) modInt MAX => #getRange(BM, ADDR, WIDTH)
+      requires 0 <Int MAX andBool 0 <Int WIDTH
+       andBool 2 ^Int (8 *Int WIDTH) <=Int MAX
```

Comment:
Could this be expressed as `#getRange(BM, ADDR, WIDTH) <=Int MAX => true requires ...`, and have the regular rules for `modInt` simplify? I've been trying to add things like that before, but I'm not sure simplifications get applied in checking side conditions.

Reply:
I would like to be able to simplify this just using the `modInt` rule, but I think you're correct that simplification rules don't apply when evaluating side conditions or something, because I've had issues getting some cases like this to simplify.

Reply:
Okay, I'll make an issue about that, since I've had the same experience, and simplifications applying when evaluating side conditions is a pretty important part of being able to compose lemmas.

Reply:
I've sucessfully removed this lemma now.